orttraining/orttraining/training_api/optimizer.cc

```diff
@@ -48,59 +49,48 @@ Status Optimizer::GenerateMomentumNamedStates() {
-    // Collect all the inputs based on the gradient names order.
-    for (size_t i = 5; i < input_names_.size(); i++) {
-      std::string param_name;
-      if (utils::GetParamNameFromGradient(input_names_[i], param_name)) {
-        const auto named_parameter_it = named_parameters_.find(param_name);
-        ORT_ENFORCE(named_parameter_it != named_parameters_.end(),
-                    "Unknown param: ", param_name, " for field: ", input_names_[i]);
-
-        // Collect the gradients as ortvalues
-        grads.push_back(named_parameter_it->second->Gradient());
+    std::vector<Tensor> params, grads, first_order_moments, second_order_moments;
 
+    // Collect all the non user defined inputs from the named_parameters_.
+    for (auto& name_parameter_pair : named_parameters_) {
+      if (name_parameter_pair.second->RequiresGrad()) {
         // Collect parameters and prepare for tensorseq creation
-        auto* param_tensor = named_parameter_it->second->Data().GetMutable<Tensor>();
+        auto* param_tensor = name_parameter_pair.second->Data().GetMutable<Tensor>();
         params.emplace_back(
             Tensor(param_tensor->DataType(), param_tensor->Shape(),
                    param_tensor->MutableDataRaw(), param_tensor->Location()));
 
+        // Collect gradients and prepare for tensorseq creation
+        auto* grad_tensor = name_parameter_pair.second->Gradient().GetMutable<Tensor>();
+        grads.emplace_back(
+            Tensor(grad_tensor->DataType(), grad_tensor->Shape(),
+                   grad_tensor->MutableDataRaw(), grad_tensor->Location()));
+
         // Collect first order moments and prepare for tensorseq creation
-        auto* first_order_moment_tensor = param_named_optimizer_states.at(param_name)
+        auto* first_order_moment_tensor = param_named_optimizer_states.at(name_parameter_pair.first)
                                               .momentum_named_states.at(MOMENT_STATE_NAMES[0])
                                               .GetMutable<Tensor>();
         first_order_moments.emplace_back(
             Tensor(first_order_moment_tensor->DataType(), first_order_moment_tensor->Shape(),
                    first_order_moment_tensor->MutableDataRaw(), first_order_moment_tensor->Location()));
 
         // Collect second order moments and prepare for tensorseq creation
-        auto* second_order_moment_tensor = param_named_optimizer_states.at(param_name)
+        auto* second_order_moment_tensor = param_named_optimizer_states.at(name_parameter_pair.first)
                                                .momentum_named_states.at(MOMENT_STATE_NAMES[1])
                                                .GetMutable<Tensor>();
         second_order_moments.emplace_back(
             Tensor(second_order_moment_tensor->DataType(), second_order_moment_tensor->Shape(),
                    second_order_moment_tensor->MutableDataRaw(), second_order_moment_tensor->Location()));
-      } else {
```

Comment:
why remove this else branch?

Reply:
Earlier, we were iterating over the graph inputs. So, if we encounter an unknown user input, we report it to the user in the `else` branch. We no longer need to iterate over the user inputs since the user inputs does not define any specific tensor. We need to iterate only over the named parameters and built the gradients and moments from there.

Having said that we should add a check for any unknown user inputs. Will add that.